### PR TITLE
Add Config Observer for AccessTokenInactivityTimeout in OAuth Cluster Config 

### PR DESCRIPTION
**Pull request by vareti**
Currently, access tokens are invalidated only when their token age expires. This PR wires in the bits needed to expire the token based on user inactivity. 
The config observer observes if `AccessTokenInactivityTimeoutSeconds` field is set OAuth cluster config. If timeout is set to non-zero value, it appends the below config to `KubeAPIServerConfig`.
```
oauthConfig:
  tokenConfig:
    accessTokenInactivityTimeoutSeconds: <timeout value>
```
When `AccessTokenInactivityTimeoutSeconds` is either zero or not present in OAuth cluster config,  it removes the above the config from `KubeAPIServerConfig`.

### CODE REVIEW

**Comment by deads2k** [comment]
/hold

We need an e2e test proving the kube-apiserver actually does the right thing after reacting to this setting.  `aws-operator` is the spot for now.

File: pkg/operator/targetconfigcontroller/targetconfigcontroller.go [hunk 1]
```diff
@@ -206,9 +206,7 @@ func manageKubeAPIServerConfig(client coreclientv1.ConfigMapsGetter, recorder ev
 	configMap := resourceread.ReadConfigMapV1OrDie(v410_00_assets.MustAsset("v4.1.0/kube-apiserver/cm.yaml"))
 	defaultConfig := v410_00_assets.MustAsset("v4.1.0/config/defaultconfig.yaml")
 	configOverrides := v410_00_assets.MustAsset("v4.1.0/config/config-overrides.yaml")
-	specialMergeRules := map[string]resourcemerge.MergeFunc{
-		".oauthConfig": RemoveConfig,
-	}
+	specialMergeRules := map[string]resourcemerge.MergeFunc{}
```
Comment (sttts) [inline pkg/operator/targetconfigcontroller/targetconfigcontroller.go:209]: why this?

It would deserve its own commit with an explanation.

File: test/e2e/token_inactivity_timeout_test.go [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+		UserUID:                  string(uid),
+		AuthorizeToken:           "mJOQ7Es5l9V7WYDl0bvl3E_hRjnJ21ZZxXH6YZj3yeT",
+		InactivityTimeoutSeconds: 300,
+	}
+
+	for _, accessToken := range []*oauthapi.OAuthAccessToken{shortLivedOAuthAccessToken, longLivedAccessToken} {
+		_, err = oauthClientClient.OAuthAccessTokens().Create(context.Background(), accessToken, metav1.CreateOptions{})
+		require.NoError(t, err)
+		defer func(name string) {
+			if err := oauthClientClient.OAuthAccessTokens().Delete(context.Background(), name, metav1.DeleteOptions{}); err != nil {
```
Comment (sttts) [inline test/e2e/token_inactivity_timeout_test.go:95]: nit: context.TODO()

File: test/e2e/token_inactivity_timeout_test.go [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+	resp, err := httpTransport.RoundTrip(req)
+	require.NoError(t, err)
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		t.Fatalf("short lived access token should work. Recieved status %v", resp.Status)
+	}
+
+	// Wait for token to expire.
+	fmt.Println("Waiting for short lived access token to expire due to inactivity")
```
Comment (sttts) [inline test/e2e/token_inactivity_timeout_test.go:126]: t.Log

File: test/e2e/token_inactivity_timeout_test.go [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+	// Nanosecond delta make sure that test does not hit edge cases.
+	time.Sleep(300*time.Second + time.Nanosecond)
```
Comment (sttts) [inline test/e2e/token_inactivity_timeout_test.go:128]: This `time.Nanosecond` won't make the difference. How often does the timeout controller run within kube-apiserver? Would add at least one iteration duration here.

File: test/e2e/token_inactivity_timeout_test.go [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+
+	resp, err = httpTransport.RoundTrip(req)
+	require.NoError(t, err)
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusOK {
```
Comment (sttts) [inline test/e2e/token_inactivity_timeout_test.go:134]: would be more precise here with the retun code. We would get 401 here?

File: test/e2e/token_inactivity_timeout_test.go [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+
+	resp, err = httpTransport.RoundTrip(req)
+	require.NoError(t, err)
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		t.Fatalf("long lived access token should still work. Recieved status %v", resp.Status)
+	}
+
+	fmt.Println("wait to check if long lived token works after inactivity timeout configured previously")
```
Comment (sttts) [inline test/e2e/token_inactivity_timeout_test.go:156]: you would need another short lived token here IMO. Checking the long lived one does not make much sense. It won't expire in any case.

File: test/e2e/token_inactivity_timeout_test.go [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+	}
+}
+
+func updateOAuthServerClusterConfigSpec(t *testing.T, client *configclient.ConfigV1Client, duration *metav1.Duration) {
+	oauthConfig, err := client.OAuths().Get(context.Background(), "cluster", metav1.GetOptions{})
+	require.NoError(t, err)
+
+	oauthConfig.Spec.TokenConfig.AccessTokenInactivityTimeout = duration
+
+	oauthConfig, err = client.OAuths().Update(context.Background(), oauthConfig, metav1.UpdateOptions{})
```
Comment (sttts) [inline test/e2e/token_inactivity_timeout_test.go:175]: maybe put a wait.Retry around.

File: pkg/operator/targetconfigcontroller/targetconfigcontroller.go (hunk 1 above)
Comment (stlaz) [reply pkg/operator/targetconfigcontroller/targetconfigcontroller.go:209]: The observed config is going to appear in this path. Boy do I hope that there are no side effects to this in KAS and that we cleaned all of the code that should have been dead in 4.x.

File: test/e2e/token_inactivity_timeout_test.go (hunk 4 above)
Comment (vareti) [reply test/e2e/token_inactivity_timeout_test.go:128]: The timeout controller runs every 100s. But if a token is past its validity, it expires the token instantaneously rather than waiting for next interval. https://github.com/openshift/kubernetes/blob/4d82d07bf0c5372c069fbc5dd60baa9eed3fac09/openshift-kube-apiserver/authentication/oauth/timeoutvalidator.go#L218-L230

File: test/e2e/token_inactivity_timeout_test.go [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+	req, err := http.NewRequest(http.MethodGet, kubeConfig.Host+"/apis/user.openshift.io/v1/users/~", &bytes.Buffer{})
+	require.NoError(t, err)
+	req.Header.Set("Authorization", "Bearer "+shortLivedOAuthAccessToken.Name)
+
+	resp, err := httpTransport.RoundTrip(req)
+	require.NoError(t, err)
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		t.Fatalf("short lived access token should work. Recieved status %v", resp.Status)
```
[the remark this answers is not in the document]
Comment (vareti) [reply test/e2e/token_inactivity_timeout_test.go:122]: done.

File: test/e2e/token_inactivity_timeout_test.go [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+func TestTokenInactivityTimeout(t *testing.T) {
+	kubeConfig, err := test.NewClientConfigForTest()
+	require.NoError(t, err)
+
+	userClient := userclient.NewForConfigOrDie(kubeConfig)
+	oauthClientClient := oauthclient.NewForConfigOrDie(kubeConfig)
+	configClient := configclient.NewForConfigOrDie(kubeConfig)
+
+	// Update OAuth cluster config with token inactivity timeout
+	updateOAuthServerClusterConfigSpec(t, configClient, &metav1.Duration{Duration: 300 * time.Second})
```
[the remark this answers is not in the document]
Comment (vareti) [reply test/e2e/token_inactivity_timeout_test.go:39]: done

File: test/e2e/token_inactivity_timeout_test.go (hunk 7 above)
Comment (vareti) [reply test/e2e/token_inactivity_timeout_test.go:175]: added retry mechanism

File: test/e2e/token_inactivity_timeout_test.go (hunk 6 above)
Comment (vareti) [reply test/e2e/token_inactivity_timeout_test.go:156]: is it the name that is causing confusion? the objective here is to that token is still valid after inactive period even with timeout validator disabled. any token that lives more than this duration will work in this case

File: test/e2e/token_inactivity_timeout_test.go (hunk 5 above)
Comment (vareti) [reply test/e2e/token_inactivity_timeout_test.go:134]: yes. 401. update the code accordingly

File: test/e2e/token_inactivity_timeout_test.go (hunk 3 above)
Comment (vareti) [reply test/e2e/token_inactivity_timeout_test.go:126]: done

File: test/e2e/token_inactivity_timeout_test.go (hunk 2 above)
Comment (vareti) [reply test/e2e/token_inactivity_timeout_test.go:95]: done

File: test/e2e/token_inactivity_timeout_test.go [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+		},
+	}
+
+	req, err := http.NewRequest(http.MethodGet, kubeConfig.Host+"/apis/user.openshift.io/v1/users/~", &bytes.Buffer{})
+	require.NoError(t, err)
+	req.Header.Set("Authorization", "Bearer "+shortLivedOAuthAccessToken.Name)
+
+	resp, err := httpTransport.RoundTrip(req)
+	require.NoError(t, err)
+	defer resp.Body.Close()
```
[the remark this answers is not in the document]
Comment (vareti) [reply test/e2e/token_inactivity_timeout_test.go:119]: good to know about this. thanks!

File: pkg/operator/targetconfigcontroller/targetconfigcontroller.go [hunk 11]
```diff
@@ -206,9 +206,7 @@ func manageKubeAPIServerConfig(client coreclientv1.ConfigMapsGetter, recorder ev
 	configMap := resourceread.ReadConfigMapV1OrDie(v410_00_assets.MustAsset("v4.1.0/kube-apiserver/cm.yaml"))
 	defaultConfig := v410_00_assets.MustAsset("v4.1.0/config/defaultconfig.yaml")
 	configOverrides := v410_00_assets.MustAsset("v4.1.0/config/config-overrides.yaml")
-	specialMergeRules := map[string]resourcemerge.MergeFunc{
-		".oauthConfig": RemoveConfig,
```
Comment (sttts) [inline pkg/operator/targetconfigcontroller/targetconfigcontroller.go:210]: add to the commit message why it can go now.

File: test/e2e/token_inactivity_timeout_test.go [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+	t.Log("Waiting for short lived access token to expire due to inactivity")
+	// Nanosecond delta make sure that test does not hit edge cases.
+	time.Sleep(300*time.Second + time.Nanosecond)
+
+	resp, err = httpTransport.RoundTrip(req)
+	require.NoError(t, err)
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusUnauthorized {
+		t.Fatalf("short lived access token should not work. Recieved status %v", resp.Status)
```
Comment (sttts) [inline test/e2e/token_inactivity_timeout_test.go:135]: nit: typo "received"

File: test/e2e/token_inactivity_timeout_test.go (hunk 6 above)
Comment (sttts) [reply test/e2e/token_inactivity_timeout_test.go]: Am a little confused. We document in the API https://github.com/openshift/api/blob/master/config/v1/types_oauth.go#L55 that tokens expire no-matter-what after the configured timeout. How does the set timeout value in the access token play into this? From the API docs of `TokenConfig.AccessTokenInactivityTimeout`, even the long-lived token should expire. What do I miss?

**Comment by sttts** [comment]
I would expect these tests:
1. tokens with and without per-token inactivity timeouts, and unset `TokenConfig.AccessTokenInactivityTimeout` expire as defined (or don't expire for the second token without per-token inactivity timeouts).
2. set the `TokenConfig.AccessTokenInactivityTimeout`. Again try token with and without per-token inactivity timeouts. Now both expire.
3. set the per-client `AccessTokenInactivityTimeout` and check that it trumps over the `TokenConfig.AccessTokenInactivityTimeout`.
4. unset `TokenConfig.AccessTokenInactivityTimeout` and check that the per-client timeout still works.
5. unset per-client timeout and try that we are back at the behaviour of (1)

File: test/e2e/token_inactivity_timeout_test.go (hunk 6 above)
Comment (vareti) [reply test/e2e/token_inactivity_timeout_test.go]: timeout validator can expire token only if [inactivity timeout is set on the token.](https://github.com/openshift/kubernetes/blob/master/openshift-kube-apiserver/authentication/oauth/timeoutvalidator.go#L81-L92) As, oauth-server issues tokens with this timeout set, tokens issued by it will expire after inactive period. So, do you think that tokens created manually without this value must also expire?

**Comment by vareti** [comment]
> wait another 2m. First token is expired.

you meant 5min?

File: test/e2e/token_inactivity_timeout_test.go [hunk 13]
```diff
@@ -0,0 +1,336 @@
...
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+		}).DialContext,
+		TLSHandshakeTimeout: 10 * time.Second,
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: true,
+		},
+	}
+
+	req, err := http.NewRequest(http.MethodGet, kubeConfig.Host+"/apis/user.openshift.io/v1/users/~", &bytes.Buffer{})
```
Comment (sttts) [inline test/e2e/token_inactivity_timeout_test.go:327]: can you move this into the checkAccess funcs?

File: test/e2e/token_inactivity_timeout_test.go [hunk 14]
```diff
@@ -0,0 +1,336 @@
...
+	require.NoError(t, err)
+
+	timeout := time.Second * time.Duration(inactivityTimeout)
```
Comment (sttts) [inline test/e2e/token_inactivity_timeout_test.go:330]: expectedInactivityTimeout. Maybe better?

File: test/e2e/token_inactivity_timeout_test.go [hunk 15]
```diff
@@ -0,0 +1,336 @@
...
+			t.Logf("%v", err)
+		}
+	}
+}
+
+func checkTokenAccess(t *testing.T,
+	kubeConfig *rest.Config,
+	userClient *userclient.UserV1Client,
+	oauthClientClient *oauthclient.OauthV1Client,
+	inactivityTimeout, oauthClientTimeout int32,
```
Comment (sttts) [inline test/e2e/token_inactivity_timeout_test.go:258]: s/inactivityTimeout/configInactivityTimeout/

File: test/e2e/token_inactivity_timeout_test.go [hunk 16]
```diff
@@ -0,0 +1,336 @@
...
+	identityName := idpName + ":" + userName
+
+	uid, cleanup := createUser(t, userClient, userName, identityName)
+	defer cleanup()
+
+	cleanup = createIdentity(t, userClient, userName, identityName, idpName, uid)
+	defer cleanup()
+
+	cleanup = createOAuthClient(t, oauthClientClient, oauthClientName, redirectURIs, &oauthClientTimeout)
+	defer cleanup()
```
Comment (sttts) [inline test/e2e/token_inactivity_timeout_test.go:275]: 3 times cleanup?

File: test/e2e/token_inactivity_timeout_test.go [hunk 17]
```diff
@@ -0,0 +1,336 @@
...
+	}
+}
+
+func checkTokenAccess(t *testing.T,
+	kubeConfig *rest.Config,
+	userClient *userclient.UserV1Client,
+	oauthClientClient *oauthclient.OauthV1Client,
+	inactivityTimeout, oauthClientTimeout int32,
+	prefix string,
+	checkAccess func(*testing.T, *http.Transport, *http.Request, *oauthapi.OAuthAccessToken, *oauthapi.OAuthAccessToken, time.Duration)) {
```
Comment (sttts) [inline test/e2e/token_inactivity_timeout_test.go:260]: Usually we prefix them with `test` instead of `check`.

File: test/e2e/token_inactivity_timeout_test.go [hunk 18]
```diff
@@ -0,0 +1,336 @@
...
+	updateOAuthConfigInactivityTimeout(t, configClient, nil)
+	// Wait for kube-apiserver to start the new roll out.
+	test.WaitForKubeAPIServerStartProgressing(t, configClient)
+	// Wait until all kube-apiserver pods are restarted with new config.
+	test.WaitForKubeAPIServerClusterOperatorAvailableNotProgressingNotDegraded(t, configClient)
+	// No OAuthClient timeout and no OAuth config timeout.
+	checkTokenAccess(t, kubeConfig, userClient, oauthClientClient, 10, 0, "without-inactivity-timeout", tokenCheck1)
+}
+
+func updateOAuthConfigInactivityTimeout(t *testing.T, client *configclient.ConfigV1Client, duration *metav1.Duration) {
```
Comment (sttts) [inline test/e2e/token_inactivity_timeout_test.go:156]: move all helpers to the bottom. That's usually done in Go.

File: test/e2e/token_inactivity_timeout_test.go (hunk 17 above)
Comment (sttts) [reply test/e2e/token_inactivity_timeout_test.go:260]: How about having arbitrary many test funcs here? That would allow you to give them nice names further up.

Like:
```
checkTokenAccess(t, kubeConfig, userClient, oauthClientClient, 10, 0, "without-inactivity-timeout",
   tokenWithInactivityTimeoutExpired,
   tokenWithoutInactivityTimeoutWorks,
)
```

File: test/e2e/token_inactivity_timeout_test.go [hunk 19]
```diff
@@ -0,0 +1,336 @@
...
+	require.NoError(t, err)
+
+	userClient := userclient.NewForConfigOrDie(kubeConfig)
+	oauthClientClient := oauthclient.NewForConfigOrDie(kubeConfig)
+	configClient := configclient.NewForConfigOrDie(kubeConfig)
+
+	// This checks that
+	// 1. token with timeout does not work after timeout
+	// 2. token without timeout works anytime
+	tokenCheck1 := func(t *testing.T, transport *http.Transport, req *http.Request, tokenWithTimeout, tokenWithoutTimeout *oauthapi.OAuthAccessToken, timeout time.Duration) {
```
Comment (sttts) [inline test/e2e/token_inactivity_timeout_test.go:43]: give those two a better names, compare https://github.com/openshift/cluster-kube-apiserver-operator/pull/874/files?file-filters%5B%5D=.go&file-filters%5B%5D=.md&file-filters%5B%5D=.mod&file-filters%5B%5D=.txt&file-filters%5B%5D=.yml&file-filters%5B%5D=No+extension&file-filters%5B%5D=dotfile#r463130000. I.e. split these up into two checks:

- tokenWithInactivityTimeoutExpired
- tokenWithoutInactivityTimeoutWorks

File: test/e2e/token_inactivity_timeout_test.go [hunk 20]
```diff
@@ -0,0 +1,336 @@
...
+			t.Fatalf("token with out timeout should work. Received status %q", resp.Status)
+		}
+	}
+
+	// No OAuthClient timeout and no OAuth config timeout.
+	checkTokenAccess(t, kubeConfig, userClient, oauthClientClient, 10, 0, "without-inactivity-timeout", tokenCheck1)
+
+	// Update OAuth cluster config with token inactivity timeout.
+	updateOAuthConfigInactivityTimeout(t, configClient, &metav1.Duration{Duration: 600 * time.Second})
+	// Wait for kube-apiserver pods to start progressing.
```
Comment (sttts) [inline test/e2e/token_inactivity_timeout_test.go:136]: these comments disturb reading flow more than they help. They don't add anything that's not clear from the func name.

File: test/e2e/token_inactivity_timeout_test.go [hunk 21]
```diff
@@ -0,0 +1,336 @@
...
+	test.WaitForKubeAPIServerStartProgressing(t, configClient)
+	// Wait until all kube-apiserver pods are restarted with new config.
+	test.WaitForKubeAPIServerClusterOperatorAvailableNotProgressingNotDegraded(t, configClient)
+
+	// With only OAuth config timeout.
+	checkTokenAccess(t, kubeConfig, userClient, oauthClientClient, 600, 0, "with-inactivity-timeout", tokenCheck2)
```
Comment (sttts) [inline test/e2e/token_inactivity_timeout_test.go:142]: run these as sub-tests:
```
t.Run("with-inactivity-timeout", func(t*testing.T) {
  checkTokenAccess(t, kubeConfig, userClient, oauthClientClient, 600, 0, tokenCheck2)
}
```

File: test/e2e/token_inactivity_timeout_test.go [hunk 22]
```diff
@@ -0,0 +1,284 @@
...
+	t.Run("unset-inactivity-timeout-client-timeout", func(t *testing.T) {
+		checkTokenAccess(t, userClient, oauthClientClient, &configInactivityTimeout, nil, testTokenTimeoutScenarios)
+	})
+
+}
+
+func checkTokenAccess(t *testing.T,
+	userClient *userclient.UserV1Client,
+	oauthClientClient *oauthclient.OauthV1Client,
+	configInactivityTimeout, oauthClientTimeout *int32,
```
Comment (sttts) [inline test/e2e/token_inactivity_timeout_test.go:121]: why is configInactivityTimeout a pointer?

File: test/e2e/token_inactivity_timeout_test.go [hunk 23]
```diff
@@ -0,0 +1,284 @@
...
+			Name: prefix + "token-with-timeout",
+		},
+		ClientName:               oauthClientName,
+		ExpiresIn:                defaultAccessTokenMaxAgeSeconds,
+		Scopes:                   []string{"user:full"},
+		RedirectURI:              redirectURIs[0],
+		UserName:                 userName,
+		UserUID:                  string(uid),
+		AuthorizeToken:           "mJOQ7Es5l9V7WYDl0bvl3E_hRjnJ21ZZxXH6YZj3yeS",
+		InactivityTimeoutSeconds: expectedInactivityTimeout,
```
Comment (sttts) [inline test/e2e/token_inactivity_timeout_test.go:158]: this must be much shorter I think. Would expect something <<120s, such that the use of the token is pushing the timeout beyond 120s. Otherwise the test after 120s will always work, even if `configInactivityTimeout`/`oauthClientTimeout` are not wired correctly into kube-apiserver.

Comment (sttts) [reply test/e2e/token_inactivity_timeout_test.go:158]: Make it 60s and we are fine.

File: test/e2e/token_inactivity_timeout_test.go (hunk 22 above)
Comment (vareti) [reply test/e2e/token_inactivity_timeout_test.go]: no specific reason. I guess I was just lazy to declare variable of `int32` type

File: test/e2e/token_inactivity_timeout_test.go (hunk 23 above)
Comment (vareti) [reply test/e2e/token_inactivity_timeout_test.go:158]: made it 60s now.

**Comment by vareti** [comment]
/hold cancel

@sttts addressed your last two comments. please take a look. thanks.

**Comment by sttts** [comment]
Terraform error.

/retest

**Comment by sttts** [comment]
AWS lease error.

/retest
